pkg/distributor/distributor.go

```diff
@@ -201,14 +203,47 @@ func (d *Distributor) removeStaleIngesterClients() {
 
 		// Do the gRPC closing in the background since it might take a while and
 		// we're holding a mutex.
-		go func(addr string, closer io.Closer) {
-			if err := closer.Close(); err != nil {
-				level.Error(util.Logger).Log("msg", "error closing connection to ingester", "ingester", addr, "err", err)
-			}
-		}(addr, client.(io.Closer))
+		go closeClient(addr, client.(io.Closer))
+	}
+}
+
+func closeClient(addr string, closer io.Closer) {
+	if err := closer.Close(); err != nil {
+		level.Error(util.Logger).Log("msg", "error closing connection to ingester", "ingester", addr, "err", err)
+	}
+}
+
+func (d *Distributor) healthCheckAndRemoveIngesters() {
+	ingesters := d.ring.GetAll()
+	for _, ingester := range ingesters {
+		d.healthCheckAndRemoveIngester(ingester)
+	}
+}
+
+func (d *Distributor) healthCheckAndRemoveIngester(ingester *ring.IngesterDesc) {
+	client, err := d.getClientFor(ingester)
+	if err != nil {
+		d.removeClientFor(ingester, err)
+	}
```

Comment:
if err != nil, client == nil and the rest of this function will panic.

Reply:
Yeah I think it err != nil you don't need to remove the client either, you should just log and return (or return an error).